lib/attrsets.nix

```diff
@@ -114,7 +114,7 @@ rec {
        => [1 2]
   */
   catAttrs = builtins.catAttrs or
-    (attr: l: concatLists (map (s: if s ? ${attr} then [s.${attr}] else []) l));
+    (attr: concatMap (s: if s ? ${attr} then [s.${attr}] else []));
```

Comment:
the catAttrs builtin was added in 2014. at this point it might make sense to just remove the fallback altogether and move this binding to trivial.nix instead.

Reply:
right, concatMap was added in the release 2.1 which is after 2014, so this wouldn't actually work. I was hoping to save an env allocation here, but I should of checked this beforehand.

Reply:
nix could be doing this kind of env elision automatically. we'll see if we can fit it into our optimization system and check whether it makes a difference on nixpkgs :)

Reply:
hmm, I saw roberth mention that currying is still preferable in https://github.com/NixOS/nixpkgs/pull/147077#discussion_r767178266, so I presume it hasn't been optimized yet. However, I haven't delve into the Nix codebase yet so I wouldn't know.

Reply:
yeah, every lambda requires one env allocation at this point. it's reasonably easy to elide envs for lambdas that don't use their parameters, but it's currently really hard to squash envs for curried functions. :( the analysis required to do this may be more expensive than just ignoring the problem.